src/stream/src/common/table/state_table.rs

```diff
@@ -698,25 +700,43 @@ where
         // Compute output indices
         let (_, output_indices) = find_columns_by_ids(&columns[..], &output_column_ids);
 
-        // Get clean watermark PK index using the helper method
-        let clean_watermark_index_in_pk: Option<i32> = table_catalog
-            .get_clean_watermark_index_in_pk_compat()
-            .map(|idx| idx as i32);
+        let clean_watermark_indices = table_catalog.get_clean_watermark_column_indices();
+        if clean_watermark_indices.len() > 1 {
+            unimplemented!("multiple clean watermark columns are not supported yet")
+        }
+        let clean_watermark_index = clean_watermark_indices.first().map(|&i| i as usize);
+
+        let watermark_serde = clean_watermark_index.map(|idx| {
+            let pk_idx = pk_indices.iter().position(|&i| i == idx);
+            let (watermark_serde, watermark_serde_type) = match pk_idx {
+                Some(0) => (pk_serde.index(0).into_owned(), WatermarkSerdeType::PkPrefix),
+                Some(pk_idx) => (
+                    pk_serde.index(pk_idx).into_owned(),
+                    WatermarkSerdeType::NonPkPrefix,
+                ),
+                None => (
+                    OrderedRowSerde::new(
+                        vec![data_types[idx].clone()],
+                        vec![OrderType::ascending()],
+                    ),
+                    // TODO(ttl): may introduce a new type for watermark not in pk.
+                    WatermarkSerdeType::NonPkPrefix,
+                ),
+            };
+            (watermark_serde, watermark_serde_type)
+        });
 
         // Restore persisted table watermark.
-        let watermark_serde = if pk_indices.is_empty() {
```

Comment:
I did once encounter `pk_indices.is_empty()` in e2e. Do you know what could lead to it?